File: audit_duplicates.py
import json
import sys
from pathlib import Path
from collections import defaultdict

questions_file = Path(__file__).parent / 'data' / 'questions.json'

try:
    with open(questions_file, 'r', encoding='utf-8') as f:
        questions = json.load(f)
    
    total_count = len(questions) if isinstance(questions, list) else 1
    
    print(f"\nTotal questions: {total_count}\n")
    
    # Check for duplicate IDs
    print("=== DUPLICATE ID CHECK ===")
    id_map = defaultdict(list)
    
    for idx, q in enumerate(questions):
        id_map[q['id']].append(idx)
    
    duplicate_ids = {id_val: indices for id_val, indices in id_map.items() if len(indices) > 1}
    
    if duplicate_ids:
        print(f"❌ Found {len(duplicate_ids)} duplicate ID(s):")
        for id_val, indices in sorted(duplicate_ids.items()):
            print(f"  ID '{id_val}' appears {len(indices)} times at indices: {indices}")
    else:
        print("✓ No duplicate IDs found\n")
    
    # Check for duplicate question text
    print("\n=== DUPLICATE QUESTION TEXT CHECK ===")
    question_map = defaultdict(list)
    
    for idx, q in enumerate(questions):
        question_text = q['question']
        question_map[question_text].append(idx)
    
    duplicate_questions = {q_text: indices for q_text, indices in question_map.items() if len(indices) > 1}
    
    if duplicate_questions:
        print(f"❌ Found {len(duplicate_questions)} duplicate question(s):")
        for question_text, indices in sorted(duplicate_questions.items()):
            print(f"\n  Question appears {len(indices)} times:")
            print(f"  \"{question_text[:80]}{'...' if len(question_text) > 80 else ''}\"")
            for idx in indices:
                print(f"    - Index {idx}: ID \"{questions[idx]['id']}\"")
    else:
        print("✓ No duplicate question text found\n")
    
    # Summary
    print("\n=== SUMMARY ===")
    print(f"Total questions: {total_count}")
    print(f"Duplicate IDs: {len(duplicate_ids)}")
    print(f"Duplicate questions: {len(duplicate_questions)}")
    
    if len(duplicate_ids) == 0 and len(duplicate_questions) == 0:
        print("\n✓ Audit complete: No duplicates found!\n")
    else:
        print("\n⚠️ Audit complete: Duplicates detected!\n")
    
except Exception as e:
    print(f"Error: {e}", file=sys.stderr)
    sys.exit(1)

File: audit-duplicates.js
const fs = require('fs');
const path = require('path');

const questionsFile = path.join(__dirname, 'data', 'questions.json');

try {
  const questions = JSON.parse(fs.readFileSync(questionsFile, 'utf8'));
  const totalCount = Array.isArray(questions) ? questions.length : 1;
  
  console.log(`\nTotal questions: ${totalCount}\n`);
  
  // Check for duplicate IDs
  console.log('=== DUPLICATE ID CHECK ===');
  const idMap = {};
  const duplicateIds = {};
  
  questions.forEach((q, idx) => {
    if (idMap[q.id]) {
      if (!duplicateIds[q.id]) {
        duplicateIds[q.id] = [idMap[q.id]];
      }
      duplicateIds[q.id].push(idx);
    } else {
      idMap[q.id] = idx;
    }
  });
  
  if (Object.keys(duplicateIds).length > 0) {
    console.log(`❌ Found ${Object.keys(duplicateIds).length} duplicate ID(s):`);
    Object.entries(duplicateIds).forEach(([id, indices]) => {
      console.log(`  ID "${id}" appears ${indices.length + 1} times at indices: ${indices.join(', ')}, ${Object.keys(idMap).indexOf(id) !== -1 ? idMap[id] : '?'}`);
    });
  } else {
    console.log('✓ No duplicate IDs found\n');
  }
  
  // Check for duplicate question text
  console.log('\n=== DUPLICATE QUESTION TEXT CHECK ===');
  const questionMap = {};
  const duplicateQuestions = {};
  
  questions.forEach((q, idx) => {
    const questionText = q.question;
    if (questionMap[questionText]) {
      if (!duplicateQuestions[questionText]) {
        duplicateQuestions[questionText] = [questionMap[questionText]];
      }
      duplicateQuestions[questionText].push(idx);
    } else {
      questionMap[questionText] = idx;
    }
  });
  
  if (Object.keys(duplicateQuestions).length > 0) {
    console.log(`❌ Found ${Object.keys(duplicateQuestions).length} duplicate question(s):`);
    Object.entries(duplicateQuestions).forEach(([questionText, indices]) => {
      console.log(`\n  Question appears ${indices.length + 1} times:`);
      console.log(`  "${questionText.substring(0, 80)}${questionText.length > 80 ? '...' : ''}"`);
      const allIndices = [questionMap[questionText], ...indices];
      allIndices.forEach(idx => {
        console.log(`    - Index ${idx}: ID "${questions[idx].id}"`);
      });
    });
  } else {
    console.log('✓ No duplicate question text found\n');
  }
  
  // Summary
  console.log('\n=== SUMMARY ===');
  console.log(`Total questions: ${totalCount}`);
  console.log(`Duplicate IDs: ${Object.keys(duplicateIds).length}`);
  console.log(`Duplicate questions: ${Object.keys(duplicateQuestions).length}`);
  
  if (Object.keys(duplicateIds).length === 0 && Object.keys(duplicateQuestions).length === 0) {
    console.log('\n✓ Audit complete: No duplicates found!\n');
  } else {
    console.log('\n⚠️ Audit complete: Duplicates detected!\n');
  }
  
} catch (error) {
  console.error('Error:', error.message);
  process.exit(1);
}
